docs/languages/csharp-iterators.md

```diff
@@ -0,0 +1,247 @@
+# Introduction
```

Comment:
I think this paragraph can be broken up to really bring home the value by enumerating (pun intended!) scenarios as a bulleted list:

* Doing something for each item in a collection
* Producing a way to enumerate a custom collection
* Extending LINQ or other collection libraries
* Writing a data pipeline which "streams" items efficiently through it

There's probably a few more, but these are areas I've seen the value add of iterators.